exonum/src/api/websocket.rs

```diff
@@ -19,58 +19,206 @@ use actix_web::ws;
 
 use rand::{rngs::ThreadRng, Rng};
 
-use std::{cell::RefCell, collections::HashMap, sync::Arc};
+use futures::Future;
 
-use crate::api::ServiceApiState;
-use crate::blockchain::Schema;
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
+};
+
+use crate::api::{
+    node::public::explorer::{TransactionHex, TransactionResponse},
+    ServiceApiState,
+};
+use crate::blockchain::{Block, Schema, TransactionResult, TxLocation};
 use crate::crypto::Hash;
+use crate::events::error::into_failure;
+use crate::explorer::TxStatus;
+use crate::messages::{Message as ExonumMessage, ProtocolMessage, RawTransaction, SignedMessage};
+
+use exonum_merkledb::{IndexAccess, ListProof, Snapshot};
+
+/// Message, coming from websocket connection.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(tag = "type", content = "payload", rename_all = "kebab-case")]
+enum IncomingMessage {
+    /// Set subscription for websocket connection.
+    SetSubscriptions(Vec<SubscriptionType>),
+    /// Send transaction to blockchain.
+    Transaction(TransactionHex),
+}
+
+/// Subscription type (new blocks or committed transactions).
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Hash, Clone, PartialOrd, Ord)]
+#[serde(tag = "type", rename_all = "kebab-case")]
+pub enum SubscriptionType {
+    /// Subscription to nothing.
+    None,
+    /// Subscription on new blocks.
+    Blocks,
+    /// Subscription on committed transactions.
+    Transactions {
+        /// Optional filter for subscription.
+        filter: Option<TransactionFilter>,
+    },
+}
+
+/// Describe filter for transactions by ID of service and (optionally)
+/// transaction type in service.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Hash, Clone, PartialOrd, Ord)]
+pub struct TransactionFilter {
+    /// ID of service.
+    pub service_id: u16,
+    /// Optional ID of transaction in service (if not set, all transaction of service will be sent).
+    pub message_id: Option<u16>,
+}
+
+impl TransactionFilter {
+    /// Create new transaction filter.
+    pub fn new(service_id: u16, message_id: Option<u16>) -> Self {
+        Self {
+            service_id,
+            message_id,
+        }
+    }
+}
+
+/// Information about a particular transaction in the blockchain (without transaction content).
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct TransactionInfo {
+    tx_hash: Hash,
+    /// ID of service.
+    pub service_id: u16,
+    /// ID of transaction in service.
+    pub message_id: u16,
+    #[serde(with = "TxStatus")]
+    status: TransactionResult,
+    location: TxLocation,
+    proof: ListProof<Hash>,
+}
+
+impl TransactionInfo {
+    fn new<T>(schema: &Schema<T>, tx_hash: &Hash) -> Self
+    where
+        T: AsRef<dyn Snapshot> + IndexAccess,
+    {
+        let tx = schema.transactions().get(tx_hash).unwrap();
```

Comment:
@Flowneee What if we don't have a transaction with hash `tx_hash` in the blockchain ? What do you think about to return a `Result` or `Option` here ?

Reply:
@aleksuss *Currently* there is always will be a transaction, because the only place where `TransactionInfo` called is [here](https://github.com/exonum/exonum/pull/1335/files/5b833cf0865ec25bf7e54a50dc563f48eea4521f#diff-581fbced55c30b0ca3a20d76b965c630R277) (`tx_hash` in that case is always already committed transaction). Generally yes, probably good idea. And I even think that method should be not `new`, but something like `new_committed` to indicate that we want info about committed transaction, and this method should return `None` instead of each `.unwrap()`. Another option is to rename type to something like `CommittedTransactionInfo`. What do you say?

Reply:
@Flowneee I've got your point and it's true. But `unwrap()` hurts my eyes in this place. I'd prefer a case with `None` instead of each `.unwrap()` but also you can use `.expect()` with a message like ("BUG. We don't have committed transaction in `transactions` index with hash: {:?}, tx_hash). Because the transaction is really MUST BE in the index. So. It's up to you.

Reply:
Okay, done. `TransactionInfo` renamed to `CommittedTransactionSummary`, added logging of situation, when summary cannot be built.

Reply:
@alekseysidorov @aleksuss @popzxc Can you check it again?